finetune/model.py

```diff
@@ -101,17 +99,54 @@ def fp16_variable_getter(
     )
 
 
-def get_variable_getter(estimator_mode, features, fp16_predict):
-    if estimator_mode == tf.estimator.ModeKeys.PREDICT and fp16_predict:
+def mp_variable_getter(
+    getter,
+    name,
+    shape=None,
+    dtype=None,
+    initializer=None,
+    regularizer=None,
+    trainable=True,
+    *args,
+    **kwargs
+):
+    var = getter(
+        name,
+        shape,
+        dtype=tf.float32 if dtype in [tf.float16, tf.float32] else dtype,
+        initializer=initializer,
+        regularizer=regularizer,
+        trainable=trainable,
+        *args,
+        **kwargs
+    )
+    if var.dtype == tf.float32:
+        var = tf.cast(var, tf.float16)
+    return var
+
+
+def get_variable_getter(estimator_mode, features, fp16_predict, mixed_precision):
+    if estimator_mode == tf.estimator.ModeKeys.TRAIN and mixed_precision:
+        custom_getter = mp_variable_getter
+        features = tf.nest.map_structure(
+            lambda feat: tf.cast(feat, tf.float16)
+            if feat.dtype == tf.float32
+            else feat,
+            features,
+        )
+    elif estimator_mode == tf.estimator.ModeKeys.PREDICT and fp16_predict:
         custom_getter = fp16_variable_getter
         features = tf.nest.map_structure(
-            lambda feat: tf.cast(feat, tf.float16) if feat.dtype == tf.float32 else feat,
-            features
+            lambda feat: tf.cast(feat, tf.float16)
+            if feat.dtype == tf.float32
```

Comment:
there is no need to worry about tf.float64 getting out of experimental and sneaking through this is there?

Reply:
So this is intentionally keeping anything that is fp64 as fp64?  Do we have anything that is / should be fp64?

Reply:
Is it typical to make prediction pure fp16 or is there potential benefit to mixed precision inference?

Reply:
If anything is float64 I think we should assume it is intentional and would already be cast before being used in the model. I don't think we have anything that generates fp64 input. Pure float 16 would be slightly preferable over mixed at inference just due to the size of the weights.